hackage-db/src/Distribution/Hackage/DB/Path.hs

```diff
@@ -15,10 +15,37 @@ import Distribution.Hackage.DB.Errors
 
 import Control.Exception
 import System.Directory
+import System.Environment (lookupEnv)
 import System.FilePath
 
+-- |
+-- Determines the /state/ directory (which e.g. holds the hackage tarball)
+-- cabal-install uses via the following logic:
+--
+-- 1. If the @CABAL_DIR@ environment variable is set, its content is used as the
+--    cabal state directory
+-- 2. If @~/.cabal@ (see 'getAppUserDataDirectory') exists, use that.
+-- 3. Otherwise, use @${XDG_CACHE_HOME}/cabal@ (see @'getXdgDirectory' 'XdgCache'@)
+--    which is the new directory cabal-install can use starting with
+--    version @3.10.*@.
+--
+-- This logic is mostly equivalent to what upstream cabal-install is
+-- [doing](https://github.com/haskell/cabal/blob/0ed12188525335ac9759dc957d49979ab09382a1/cabal-install/src/Distribution/Client/Config.hs#L594-L610)
+-- with the following exception:
+-- The state directory can freely be configured to use a different location
+-- in the cabal-install configuration file. hackage-db doesn't parse this
+-- configuration file, so differing state directories are ignored.
 cabalStateDir :: IO FilePath
-cabalStateDir = getAppUserDataDirectory "cabal"
+cabalStateDir = do
+  envCabal <- lookupEnv "CABAL_DIR"
+  dotCabal <- getAppUserDataDirectory "cabal"
+  dotCabalExists <- doesDirectoryExist dotCabal
+  case envCabal of
+    Just dir -> pure dir
+    Nothing ->
+      if dotCabalExists
+      then pure dotCabal
+      else getXdgDirectory XdgCache "cabal"
```

Comment:
```suggestion
    Nothing | dotCabalExists -> pure dotCabal
    Nothing -> getXdgDirectory XdgCache "cabal"
```
Really unimportant. Just suggesting it to learn whether you also like it better and if not, why not.

Reply:
This was copy-pasted from the cabal source code which doesn't seem to use guards much.